src/schedule/kat_key_schedule.rs

```diff
@@ -0,0 +1,282 @@
+//! # Known Answer Tests for the key schedule
+//!
+//! See https://github.com/mlswg/mls-implementations/blob/master/test-vectors.md
+//! for more description on the test vectors.
+//!
+//! If values are not present, they are encoded as empty strings.
+
+use std::convert::TryFrom;
+
+use crate::{
+    ciphersuite::{Ciphersuite, CiphersuiteName},
+    codec::Codec,
+    config::Config,
+    group::{GroupContext, GroupEpoch, GroupId},
+    schedule::{EpochSecrets, InitSecret, JoinerSecret, KeySchedule, WelcomeSecret},
+    test_util::{bytes_to_hex, hex_to_bytes, read, write},
+    utils::randombytes,
+};
+
+use hpke::HPKEKeyPair;
+use serde::{self, Deserialize, Serialize};
+
+use super::CommitSecret;
+
+#[derive(Serialize, Deserialize, Debug, Clone, Default)]
+struct Epoch {
+    // Chosen by the generator
+    tree_hash: String,
+    commit_secret: String,
+    // XXX: PSK is not supported in OpenMLS yet #141
+    psk_secret: String,
+    confirmed_transcript_hash: String,
+
+    // Computed values
+    group_context: String,
+    joiner_secret: String,
+    welcome_secret: String,
+    init_secret: String,
+    sender_data_secret: String,
+    encryption_secret: String,
+    exporter_secret: String,
+    authentication_secret: String,
+    external_secret: String,
+    confirmation_key: String,
+    membership_key: String,
+    resumption_secret: String,
+
+    external_pub: String, // TLS serialized HPKEPublicKey
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone, Default)]
+struct KeyScheduleTestVector {
+    cipher_suite: u16,
+    group_id: String,
+    initial_init_secret: String,
+    epochs: Vec<Epoch>,
+}
+
+fn generate(
+    ciphersuite: &'static Ciphersuite,
+    init_secret: &InitSecret,
+    group_id: &[u8],
+    epoch: u64,
+) -> (
+    Vec<u8>,
+    CommitSecret,
+    JoinerSecret,
+    WelcomeSecret,
+    EpochSecrets,
+    Vec<u8>,
+    GroupContext,
+    HPKEKeyPair,
+) {
+    let tree_hash = randombytes(ciphersuite.hash_length());
+    let commit_secret = CommitSecret::random(ciphersuite.hash_length());
+    let joiner_secret = JoinerSecret::new(ciphersuite, &commit_secret, init_secret);
+    let mut key_schedule = KeySchedule::init(ciphersuite, joiner_secret.clone(), None);
+    let welcome_secret = key_schedule.welcome().unwrap();
+
+    let confirmed_transcript_hash = randombytes(ciphersuite.hash_length());
+
+    let group_context = GroupContext::new(
+        GroupId::from_slice(group_id),
+        GroupEpoch(epoch),
+        tree_hash.to_vec(),
+        confirmed_transcript_hash.clone(),
+        &[], // Extensions
```

Comment:
You could add the (one and only) extension here if you wanted. They are not supported in the MLS groups but should work in the key schedule already. You would have to tweak the tree hash however.

Reply:
It's currently not part of the test vectors. We can extend them and then add them here. But this is the definition right now and it doesn't really change the key schedule. So I'm not sure we want to add anything here.